sympolyc/src/expr.rs

```diff
@@ -0,0 +1,241 @@
+//! Multivariate polynomial
+//! See [these notes](https://hackmd.io/@dannywillems/SyHar7p5A) for more context.
+
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Formatter, Result},
+    ops::{Add, Mul},
+};
+
+use ark_ff::PrimeField;
+use num_integer::binomial;
+use o1_utils::FieldHelpers;
+
+use crate::utils::{
+    compute_all_two_factors_decomposition, naive_prime_factors, PrimeNumberGenerator,
+};
+
+/// Represents a multivariate polynomial of degree `D` in `N` variables.
+/// The natural representation is the coefficients given in the monomial basis.
+pub struct MVPoly<F: PrimeField, const N: usize, const D: usize> {
+    coeff: Vec<F>,
+    // keeping track of the indices of the monomials that are normalized
+    // to avoid recomputing them
+    normalized_indices: Vec<usize>,
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> MVPoly<F, N, D> {
+    pub fn new() -> Self {
+        let normalized_indices = Self::compute_normalized_indices();
+        MVPoly {
+            coeff: vec![F::zero(); Self::dimension()],
+            normalized_indices,
+        }
+    }
+
+    pub fn dimension() -> usize {
+        binomial(N + D, D)
+    }
+
+    pub fn from_coeffs(coeff: Vec<F>) -> Self {
+        let normalized_indices = Self::compute_normalized_indices();
+        MVPoly {
+            coeff,
+            normalized_indices,
+        }
+    }
+
+    pub fn number_of_variables(&self) -> usize {
+        N
+    }
+
+    pub fn maximum_degree(&self) -> usize {
+        D
+    }
+
+    /// Output example for N = 2 and D = 2:
+    /// ```text
+    /// - 0 -> 1
+    /// - 1 -> 2
+    /// - 2 -> 3
+    /// - 3 -> 4
+    /// - 4 -> 6
+    /// - 5 -> 9
+    /// ```
+    pub fn compute_normalized_indices() -> Vec<usize> {
+        let length = Self::dimension();
+        let mut normalized_indices = vec![1; length];
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let primes = prime_gen.get_first_nth_primes(N);
+        let max_index = primes[N - 1].checked_pow(D as u32);
+        assert!(
+            max_index.is_some(),
+            "Overflow in computing the maximum index"
+        );
+        let max_index = max_index.unwrap();
```

Comment:
can't you merge the unwrap and assert some ?

Reply:
you can also use Option::expect()

Reply:
Done in 4e7b0f2c96dd2c5ac1606d36a61ff08f1c90d76d